chore: Adds a bus event to insert text at cursor in editor

nithindavid opened the pull request:
## Type of change
This PR will add a bus event to insert text or markdown into the editor at the current cursor position. Also refactors to use the new function for existing methods 



- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.


## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

nithindavid commented:
@iamsivin @scmmishra @pranavrajs I have added the test cases to the code walkthrough here - https://www.loom.com/share/60aa9a8603764d509009b2da52d51dd7

scmmishra commented on app/javascript/dashboard/components/widgets/WootWriter/Editor.vue:
```diff
@@ -1,5 +1,5 @@
 <template>
-  <div class="editor-root">
+  <div class="editor-root relative">
```
Is there any particular reason to make this relative? Is it to align the canned response selector?

nithindavid replied:
This is a mistake, editor root class already uses `relative`. Removing this change

scmmishra commented on app/javascript/dashboard/components/widgets/WootWriter/Editor.vue:
```diff
@@ -297,7 +298,17 @@ export default {
+
+    // BUS Event to insert text or markdown into the editor at the
+    // current cursor position.
+    // Components using this
+    // 1. SearchPopover.vue
```
Didn't understand this statement?

nithindavid replied:
I added that comment so we could track the track the components using this bus event. Let me know if this is an overkill.

scmmishra commented on app/javascript/dashboard/components/widgets/WootWriter/Editor.vue:
```diff
@@ -618,6 +607,44 @@ export default {
+      this.insertNodeIntoEditor(node, from, undefined, content);
+    },
+    insertNodeIntoEditor(node, from = 0, to = 0, content = '') {
+      if (!this.editorView) {
+        return;
+      }
+
+      // If the node is of type 'doc' and has only one child which is a paragraph,
+      // then extract its inline content to be inserted as inline.
+      if (
+        node.type.name === 'doc' &&
+        node.childCount === 1 &&
+        node.firstChild.type.name === 'paragraph'
+      ) {
+        node = node.firstChild.content;
+      } else if (content && node.textContent === content) {
+        node = this.editorView.state.schema.text(content);
+      }
+
+      let tr;
+      if (to) {
+        tr = this.editorView.state.tr
+          .replaceWith(from, to, node)
+          .insertText(` `);
+      } else {
+        tr = this.editorView.state.tr.insert(from, node);
+      }
+
+      this.state = this.editorView.state.apply(tr);
```
Would it be a good idea to move this part into a function outside the vue file, to a util? I am think that could make it easy to write unit tests. It can simply return the transaction to be applied

pranavrajs replied:
Agreed, this piece is hard to understand. It would be good to accompany it with relevant tests.

scmmishra commented on app/javascript/dashboard/components/widgets/WootWriter/Editor.vue:
```diff
@@ -458,50 +465,32 @@ export default {
-      this.emitOnChange();
+      const content = `{{${variable}}}`;
+      let node = this.editorView.state.schema.text(content);
+      const from = this.range.from;
```
```suggestion
      const { from, to } = this.range;
```

pranavrajs replied:
This should be a lint rule.

scmmishra reviewed (commented):
@nithindavid some minor comments and suggestions

nithindavid commented on app/javascript/dashboard/components/widgets/WootWriter/Editor.vue:
```diff
@@ -1,5 +1,5 @@
 <template>
-  <div class="editor-root">
+  <div class="editor-root relative">
```
```suggestion
  <div class="editor-root">
```

nithindavid reviewed (commented):
Minor comment

nithindavid commented:
@pranavrajs @scmmishra I have moved the function to a util, also added a fucntion `scrollTNode' as our current implementation does not scroll to the cursor position.

scmmishra reviewed (approved):
Code is LGTM! Assuming the functionality is working as expected from the previous loom video
